### DISCUSSION TITLE
`nameof` should support keyword type aliases

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/csharplang#7997] [Language Ideas]
`nameof` is a contextual keyword. As such, it is prone to sharp edges such as cryptic error messages when used incorrectly.

For example, `nameof(int)` gives CS1525: "[i]nvalid expression term 'int'". This is confusing as I didn't write an expression, but a type keyword. It gets worse as `nameof(int.TryParse)` works just fine (gives "TryParse"). If I want `nameof(int)` to give "Int32", I must write `nameof(Int32)`.

This also clashes with the existing lookup rules for `typeof`; `typeof(int)` returns the same as `typeof(System.Int32)`.

An open question would be: should `nameof(int)` give "int" or "Int32". I'd argue the latter, as "name of" returns ["the name of a program entity as a constant string"](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11720-nameof-expressions), and "int" does not exist at runtime. So it's not an "entity", but a keyword.

**Comment by CyrusNajmabadi** [comment]
Note: this is intentional.  The purpose of nameof is to give the actual name of something.  Note that `int` is more an `alias`, and in that regard, i'd expect it to work as `nameof` does on aliases here.  

Also, `nameof` exists entirely to link a symbol and a reference to its name *so that when one changes, the other can be updated*.  That concern doesn't exist with `Int32` as it's not going to be renamed.  If you're using `nameof` purely to reference the thing, then it's also sensible to just reference the thing in a way that clearly indicates waht it will be replaced with.

> For example, nameof(int) gives CS1525: "[i]nvalid expression term 'int'". This is confusing as I didn't write an expression, but a type keyword. 

All types *are* expressions in the language.  That's what enables writing a type on the left hand side of a `dot` for example :)

> This also clashes with the existing lookup rules for typeof; typeof(int) returns the same as typeof(System.Int32).

Sure, but `typeof` isn't `nameof`.  It's specifically intended to be about getting the end runtime type.  `nameof` is not that.  It's about allowing you to specify the name in way that isn't a string.  so you can type `nameof(Int32)` and get compile time checking, vs `"Int32"`.  In particular, we've had people ask for the exact opposite.  That `nameof(int)` return `"int"` since it's what is in the `nameof`.  

Having the rules be simple/clear/consistent is def valuable here.  When you see `nameof` you just need to look at the last actual name in the arguments, and that's what you'll get as the string value.  No question on how things will be interpreted.

As mentioned above, this is also very relevant for aliases.  if you have `using X = System.Int32;` then `nameof(X)` gives you "X" not "Int32".  This is a *very* lexical/syntactic construct.  *Not* a semantic one.  Intentionally so.
